src/node/rpc/handlerregistry.h

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/json_schema.h"
+#include "enclave/rpccontext.h"
+#include "node/certs.h"
+#include "serialization.h"
+
+#include <functional>
+#include <nlohmann/json.hpp>
+
+namespace ccf
+{
+  struct RequestArgs
+  {
+    std::shared_ptr<enclave::RpcContext> rpc_ctx;
+    Store::Tx& tx;
+    CallerId caller_id;
+    const std::string& method;
+    const nlohmann::json& params;
+  };
+
+  using HandleFunction = std::function<void(RequestArgs& args)>;
+
+  static enclave::RpcResponse make_success(nlohmann::json&& result_payload)
+  {
+    return enclave::RpcResponse{std::move(result_payload)};
+  }
+
+  static enclave::RpcResponse make_success(const nlohmann::json& result_payload)
+  {
+    return enclave::RpcResponse{result_payload};
+  }
+
+  template <typename ErrorCode>
+  static enclave::RpcResponse make_error(
+    ErrorCode code, const std::string& msg = "")
+  {
+    return enclave::RpcResponse{enclave::ErrorDetails{(int)code, msg}};
+  }
+
+  class HandlerRegistry
+  {
+  public:
+    enum ReadWrite
+    {
+      Read,
+      Write,
+      MayWrite
+    };
+
+    enum class Forwardable
+    {
+      CanForward,
+      DoNotForward
+    };
+
+    struct Handler
+    {
+      HandleFunction func;
+      ReadWrite rw;
+      nlohmann::json params_schema;
+      nlohmann::json result_schema;
+      Forwardable forwardable;
+      bool execute_locally = false;
+    };
+
+  protected:
+    std::optional<Handler> default_handler;
+    std::unordered_map<std::string, Handler> handlers;
+
+    kv::Consensus* consensus = nullptr;
+    kv::TxHistory* history = nullptr;
```

Comment:
AFAICS, both `consensus` and `history` are only needed by the child class `CommonHandlerRegistry`?

Reply:
True, but they're _set_ through this generic `HandlerRegistry` so I think its cleaner to store them here (rather than `virtual set_history(History*) {}`). This is another rough area that I will hope will polish up into something cleaner in future (briefly - only a few of the handlers need these, so it should only be visible to them).